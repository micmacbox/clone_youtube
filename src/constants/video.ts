export enum VideoType {
  default = 0,
  relatedVideos = 1,
}
