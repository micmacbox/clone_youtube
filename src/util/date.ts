import koLocale from "timeago.js/lib/lang/ko";
import { format, register } from "timeago.js";

register("ko", koLocale);

export function formatAgo(date: string, lang = "en_US") {
  return format(date, lang);
}
