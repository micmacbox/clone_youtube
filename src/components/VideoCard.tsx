import React, { FC } from "react";
import { VideoInfo } from "../@types/video";
import { formatAgo } from "../util";
import { useNavigate } from "react-router-dom";
import { VideoType } from "../constants";

const VideoCard: FC<VideoInfo> = ({ video, type }) => {
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;
  const isRelatedList = type === VideoType.relatedVideos;
  const navigate = useNavigate();
  return (
    <li
      className={isRelatedList ? "flex gap-1 m-2" : ""}
      onClick={() => {
        navigate(`/videos/watch/${video.id}`, { state: { video } });
      }}
    >
      <img
        className={isRelatedList ? "w-60 mr-2" : "w-full"}
        src={thumbnails.medium.url}
        alt={title}
      />
      <div>
        <p className="font-semibold my-2 line-clamp-2">{title}</p>
        <p className="text-sm opacity-80"> {channelTitle}</p>
        <p className="text-sm opacity-80">{formatAgo(publishedAt)}</p>
      </div>
    </li>
  );
};

export default VideoCard;
