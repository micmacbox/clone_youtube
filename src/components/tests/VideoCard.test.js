import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { Route, useLocation } from "react-router-dom";
import { formatAgo } from "../../util";
import VideoCard from "../VideoCard";
import { fakeVideo as video } from "../../test";
import { withRouter } from "../../test/utils";
import { VideoType } from "../../constants";

describe("VideoCard", () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it("renders default type correctly", () => {
    const component = renderer.create(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders relatedVideos type correctly", () => {
    const component = renderer.create(
      withRouter(
        <Route
          path="/"
          element={<VideoCard video={video} type={VideoType.relatedVideos} />}
        />
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders video item", () => {
    render(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );

    const image = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
  });

  it("navigates to detail video page with video state when cliicked", () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const card = screen.getByRole("listitem");
    userEvent.click(card);
    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
