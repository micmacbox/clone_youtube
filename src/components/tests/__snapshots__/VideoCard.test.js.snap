// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders default type correctly 1`] = `
<li
  className=""
  onClick={[Function]}
>
  <img
    alt="England vs. Italy Highlights | European Qualifiers"
    className="w-full"
    src="https://i.ytimg.com/vi/LcKoyrvqE4I/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      England vs. Italy Highlights | European Qualifiers
    </p>
    <p
      className="text-sm opacity-80"
    >
       
      TUDN USA
    </p>
    <p
      className="text-sm opacity-80"
    >
      4 days ago
    </p>
  </div>
</li>
`;

exports[`VideoCard renders relatedVideos type correctly 1`] = `
<li
  className="flex gap-1 m-2"
  onClick={[Function]}
>
  <img
    alt="England vs. Italy Highlights | European Qualifiers"
    className="w-60 mr-2"
    src="https://i.ytimg.com/vi/LcKoyrvqE4I/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      England vs. Italy Highlights | European Qualifiers
    </p>
    <p
      className="text-sm opacity-80"
    >
       
      TUDN USA
    </p>
    <p
      className="text-sm opacity-80"
    >
      4 days ago
    </p>
  </div>
</li>
`;
